lib/arm/cpu_features.c

```diff
@@ -116,4 +118,34 @@ void setup_cpu_features(void)
 	_cpu_features = features | ARM_CPU_FEATURES_KNOWN;
 }
 
+#elif defined(__aarch64__) /* not __linux__ */
+
+volatile u32 _cpu_features = 0;
+
+void setup_cpu_features(void)
+{
+	/* NEON is a mandatory part of AArch64 */
+	_cpu_features = ARM_CPU_FEATURE_NEON;
+
+	/* On FreeBSD >= 12, Linux >= 4.11 and potentially others, it is possible to use
+	   privileged system registers from userspace to check CPU feature support.
+
+	   For proper support of SoCs where different cores have different capabilities,
+	   the OS has to always report only the features supported by all cores, like FreeBSD does. */
+
+#define bits_shift(x, high, low) ((x >> low) & ((1 << (high - low + 1)) - 1))
+
+#if __FreeBSD__ >= 12
+	uint64_t isar0 = 0;
+	__asm__("mrs %0, ID_AA64ISAR0_EL1" : "=r"(isar0));
+
+	if (bits_shift(isar0, 7, 4) >= 1)
```

Comment:
Did you check the [AArch64 documentation](https://developer.arm.com/docs/ddi0595/latest/aarch64-system-registers/id_aa64isar0_el1)?  AFAICS, this check needs to be `== 2`, not `>= 1`.